### PR TITLE
Runtimeclass podapi

### PR DESCRIPTION
**Pull request by tallclair**
Adds (but doesn't implement) the RuntimeClass Pod API proposed in https://github.com/kubernetes/community/blob/master/keps/sig-node/0014-runtime-class.md

Based on the result (lazy consensus) of the alpha fields discussion on sig-architecture: https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/y9FulL9Uq6A

For https://github.com/kubernetes/features/issues/585

**Special notes for your reviewer**:

**Release note**:
```release-note
RuntimeClass is a new API resource for defining different classes of runtimes that may be used to run containers in the cluster. Pods can select a RuntimeClass to use via the RuntimeClassName field. This feature is in alpha, and the RuntimeClass feature gate must be enabled in order to use it.
```

/sig node
/kind api-change
/priority important-soon
/milestone v1.12

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -2571,6 +2571,12 @@ type PodSpec struct {
 	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
 	// +optional
 	ReadinessGates []PodReadinessGate
+	// RuntimeClassName refers to a RuntimeClass object with the same name, which should be used to
+	// run this pod.  If no RuntimeClass resource matches the named class, the pod will not be run.
+	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-node/0014-runtime-class.md
+	// This is an alpha feature and may change in the future.
+	// +optional
+	RuntimeClassName string
```
Comment (thockin) [inline pkg/apis/core/types.go]: All new optional fields must be pointers.

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -2571,6 +2571,12 @@ type PodSpec struct {
 	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
 	// +optional
 	ReadinessGates []PodReadinessGate
+	// RuntimeClassName refers to a RuntimeClass object with the same name, which should be used to
```
Comment (thockin) [inline pkg/apis/core/types.go]: Can we be more precise about which apigroup this type exists in?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -2571,6 +2571,12 @@ type PodSpec struct {
...
+	// run this pod.  If no RuntimeClass resource matches the named class, the pod will not be run.
```
Comment (thockin) [inline pkg/apis/core/types.go]: Document what happens when this is unset.

File: pkg/apis/core/types.go (hunk 1 above)
Comment (tallclair) [reply pkg/apis/core/types.go]: Done

File: pkg/apis/core/types.go (hunk 2 above)
Comment (tallclair) [reply pkg/apis/core/types.go]: Done

File: pkg/apis/core/types.go (hunk 3 above)
Comment (tallclair) [reply pkg/apis/core/types.go]: Done

**Comment by tallclair** [comment]
Thanks for the feedback @thockin 
All comments addressed.

**Comment by alexmavr** [comment]
There's a typo in the release note: `RunitmeClass`

**Pull request merged**